extensions/amp-skimlinks/0.1/link-rewriter/link-rewriter-manager.js

```diff
@@ -0,0 +1,280 @@
+        this.priorityList_);
+    // Trigger initial scan.
+    linkRewriter.onDomUpdated();
+
+    return linkRewriter;
+  }
+
+
+  /**
+   * @public
+   * Notify the LinkRewriteService that a click has happened on an anchor.
+   * This is the chance for one of the registered LinkRewriter to rewrite the
+   * url before the browser handles the click and navigates to the link url.
+   *
+   * The LinkRewriter allowed to replace the url will be chosen based on the
+   * defined priority order (in link attribute or global meta tag).
+   * If a LinkRewriter decides to not replace the url,  we will ask the next
+   * one if it wants to replace the link until one replacement is made.
+   *
+   * A "CLICK" event will also be dispatched to notify listeners that a click
+   * on an anchor has happened. This should mostly be used to send click
+   * tracking requests, handlers of this events should not
+   * mutate the anchor!
+   * @param {HTMLElement} anchor
+   * @param {string} clickType - 'click' or 'contextmenu'
+   */
+  maybeRewriteLink(anchor, clickType = 'click') {
+    const suitableLinkRewriters = this.getSuitableLinkRewritersForLink_(anchor);
+    if (suitableLinkRewriters.length) {
+      let chosenLinkRewriter = null;
+
+      // Iterate by order of priority until one of the linkRewriter
+      // replaces the link.
+      for (let i = 0; i < suitableLinkRewriters.length; i++) {
+        const hasReplaced = suitableLinkRewriters[i].rewriteAnchorUrl(anchor);
+        if (hasReplaced) {
+          chosenLinkRewriter = suitableLinkRewriters[i];
+          break;
+        }
+      }
+      const linkRewriterId = chosenLinkRewriter ? chosenLinkRewriter.id : null;
+
+      // Emit click event for analytics purposes only,
+      // anchor should not me mutated by handlers.
+      const eventData = {
+        linkRewriterId,
+        anchor,
+        clickType,
+      };
+
+      suitableLinkRewriters.forEach(linkRewriter => {
+        linkRewriter.events.send(EVENTS.CLICK, eventData);
+      });
+    }
+  }
+
+  /**
+   * @private
+   * Extract the priority list from the optional html meta tag.
+   * The meta tag should contain a whitespace separated list of
+   * LinkRewriter Ids.
+   * @param {!../../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   * @return {Array<string>}
+   */
+  getPriorityList_(ampdoc) {
+    const docInfo = Services.documentInfoForDoc(ampdoc);
+    const value = docInfo.metaTags[PRIORITY_META_TAG_NAME];
+
+    return value ? value.trim().split(/\s+/) : [];
+  }
+
+  /**
+   * @private
+   * Listen for DOM_UPDATE event.
+   * @param {*} ampPageDocument
+   */
+  installGlobalEventListener_(ampPageDocument) {
+    ampPageDocument.addEventListener(AmpEvents.DOM_UPDATE,
+        this.onDomChanged_.bind(this));
+  }
+
+  /**
+   * @private
+   * Notify all the registered LinkRewriter when new elements have been added
+   * to the page so they can re-scan the page to find potential new links to
```

Comment:
why do we need to scan the page. can we listen to click event and run the filtering at that time?

Reply:
To avoid extra redirection for the user, we try to only replace links that we know can be affiliated.
Because the way to identify if a link can be affiliated or not is asynchronous (requires request to or domain resolver API), trying to identify this at click time would be too late, we would either not have time to get the response from our server or we would need to delay the browser navigation by blocking the main thread which would obviously be bad for the user.

Instead our solution tries to identify if links can be affiliated as early as possible, in other words, when they appear on the page.

Reply:
Got you. Sounds unavoidable if you have to send over all the links on the pages to your server.